### ISSUE TITLE
Make more robust against slow cloud connection

### ISSUE THREAD

**Issue by EenerNL** [sveinse/zaptec#19]
I really like this integration but the connection doesn't look very reliable. Every minute the entities are unavailable. Do you have any idea what the cause could be?

![image](https://github.com/sveinse/zaptec/assets/114678088/52f6b834-10b3-48c3-ba72-99b7d26e6730)

**Comment by sveinse** [comment]
@EenerNL I'm not sure without any detailed logs, but perhaps the Zaptec cloud API is slow to respond and the poll for new data times out and thus becomes unavailable? Are there any errors in the logs that can point us to why?

**Comment by EenerNL** [comment]
Log text:

```
Deze fout is ontstaan door een aangepaste integratie.

Logger: custom_components.zaptec
Source: helpers/update_coordinator.py:235 
Integration: zaptec-dev (documentation) 
First occurred: 11:08:07 (6 occurrences) 
Last logged: 11:42:48

Timeout fetching zaptec-xxxx@xxx.com data
```


Running latest version, updated this morning.

**Comment by sveinse** [comment]
Thank you. I will look into it and see what can be done to make it more robust when cloud doesn't reply as fast.

**Comment by sveinse** [comment]
@EenerNL I have released a new version, `0.0.6b230924`, commit 6765670 that fixes the timeout issue. Please update and retest. I'm looking forward to hear if this fixes things for you.

**Comment by EenerNL** [comment]
Hi! Thanks for such a fast update!
I've updated to version 98ac23a.
I'll keep you updated 👍

**Comment by sveinse** [comment]
@EenerNL  FYI. Zaptec is reporting performance issues in their API, and I think this is the root cause for the problems you have observed. See https://zaptec.statuspage.io/. "we continue to observe sporadic high levels of demand that can cause slow responsiveness. If you experience a timeout, please retry the operation."  However, the timeouts in the zaptec component were way too low, and this new version should be considerable more robust

**Comment by sveinse** [comment]
@EenerNL have you had any chance to look at it?

**Comment by EenerNL** [comment]
Hi Sveinse,

The first days after update I still had the same error messages in the logs.

I just looked at the logs and found the following new log errors:

```
Deze fout is ontstaan door een aangepaste integratie.

Logger: custom_components.zaptec
Source: custom_components/zaptec/api.py:824
Integration: zaptec-dev (documentation)
First occurred: 17:37:00 (9 occurrences)
Last logged: 17:50:17

Request to https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed:
Request to https://api.zaptec.com/api/installation/xxxxxxxx-
xxxx-xxxx-xxxx-xxxxxxxxxxxx failed:
Request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/state failed:
Request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/settings failed:
Request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed:
Traceback (most recent call last):
  File "/config/custom_components/zaptec/api.py", line 786, in _request
    async with call(full_url, headers=header) as resp:
  File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 1141, in __aenter__
    self._resp = await self._coro
                 ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 560, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.11/site-packages/aiohttp/client_reqrep.py", line 899, in start
    message, payload = await protocol.read()  # type: ignore[union-attr]
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aiohttp/streams.py", line 616, in read
    await self._waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/zaptec/api.py", line 777, in _request
    async with async_timeout.timeout(API_TIMEOUT):
  File "/usr/local/lib/python3.11/site-packages/async_timeout/__init__.py", line 141, in __aexit__
    self._do_exit(exc_type)
  File "/usr/local/lib/python3.11/site-packages/async_timeout/__init__.py", line 228, in _do_exit
    raise asyncio.TimeoutError
TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/zaptec/__init__.py", line 164, in _async_update_data
    await v.state()
  File "/config/custom_components/zaptec/api.py", line 465, in state
    data = await self.circuit_info()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/zaptec/api.py", line 474, in circuit_info
    data = await self._account._request(f"circuits/{self.id}")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/zaptec/api.py", line 824, in _request
    raise RequestTimeoutError(f"Request to {full_url} failed: {err}") from err
custom_components.zaptec.api.RequestTimeoutError: Request to https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed: 
```

```
Deze fout is ontstaan door een aangepaste integratie.

Logger: custom_components.zaptec
Source: custom_components/zaptec/api.py:818
Integration: zaptec-dev (documentation)
First occurred: 17:44:15 (11 occurrences)
Last logged: 17:53:38

get request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/state failed with status 500: <ClientResponse(https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/state) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:45:33 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Vary': 'Origin', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'X-Rate-Limit-Limit': '30m', 'X-Rate-Limit-Remaining': '649', 'X-Rate-Limit-Reset': '2023-09-28T16:00:00.0000000Z', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
get request to https://api.zaptec.com/api/installation/xxxxxxxx-
xxxx-xxxx-xxxx-xxxxxxxxxxxx failed with status 500: <ClientResponse(https://api.zaptec.com/api/installation/xxxxxxxx-
xxxx-xxxx-xxxx-xxxxxxxxxxxx) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:45:39 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Vary': 'Origin', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'X-Rate-Limit-Limit': '30m', 'X-Rate-Limit-Remaining': '29888', 'X-Rate-Limit-Reset': '2023-09-28T16:00:00.0000000Z', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
get request to https://api.zaptec.com/api/chargerFirmware/installation/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed with status 500: <ClientResponse(https://api.zaptec.com/api/chargerFirmware/installation/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:45:49 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Vary': 'Origin', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'X-Rate-Limit-Limit': '30m', 'X-Rate-Limit-Remaining': '29911', 'X-Rate-Limit-Reset': '2023-09-28T16:00:00.0000000Z', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
get request to https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed with status 500: <ClientResponse(https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:48:13 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Vary': 'Origin', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'X-Rate-Limit-Limit': '30m', 'X-Rate-Limit-Remaining': '29952', 'X-Rate-Limit-Reset': '2023-09-28T16:00:00.0000000Z', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
get request to https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed with status 500: <ClientResponse(https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:53:38 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
Traceback (most recent call last):
  File "/config/custom_components/zaptec/__init__.py", line 164, in _async_update_data
    await v.state()
  File "/config/custom_components/zaptec/api.py", line 560, in state
    settings = await self._account._request(f"chargers/{self.id}/settings")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/zaptec/api.py", line 818, in _request
    raise RequestError(
custom_components.zaptec.api.RequestError: get request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/settings failed with status 500: <ClientResponse(https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/settings) [500 Internal Server Error]>
<CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:44:15 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Vary': 'Origin', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'X-Rate-Limit-Limit': '30m', 'X-Rate-Limit-Remaining': '29848', 'X-Rate-Limit-Reset': '2023-09-28T16:00:00.0000000Z', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
```

```
Deze fout is ontstaan door een aangepaste integratie.

Logger: custom_components.zaptec
Source: helpers/update_coordinator.py:233
Integration: zaptec-dev (documentation)
First occurred: 17:37:00 (13 occurrences)
Last logged: 17:53:38

Error fetching zaptec-xxxxxxxxxxxx@xxxxxxx.com data: get request to https://api.zaptec.com/api/chargerFirmware/installation/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed with status 500: <ClientResponse(https://api.zaptec.com/api/chargerFirmware/installation/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:45:49 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Vary': 'Origin', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'X-Rate-Limit-Limit': '30m', 'X-Rate-Limit-Remaining': '29911', 'X-Rate-Limit-Reset': '2023-09-28T16:00:00.0000000Z', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
Error fetching zaptec-xxxxxxxxxxxx@xxxxxxx.com data: Request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/settings failed:
Error fetching zaptec-xxxxxxxxxxxx@xxxxxxx.com data: Request to https://api.zaptec.com/api/chargers/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed:
Error fetching zaptec-xxxxxxxxxxxx@xxxxxxx.com data: Request to https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed:
Error fetching zaptec-xxxxxxxxxxxx@xxxxxxx.com data: get request to https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx failed with status 500: <ClientResponse(https://api.zaptec.com/api/circuits/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) [500 Internal Server Error]> <CIMultiDictProxy('Date': 'Thu, 28 Sep 2023 15:53:38 GMT', 'Content-Length': '0', 'Connection': 'keep-alive', 'Request-Context': 'appId=cid-v1: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains')>
```

**Comment by sveinse** [comment]
@EenerNL There is a lot of issues on the Zaptec cloud side, so many requests take a very long time. But I've attempted once more to see if it helps to increase the timeouts. Please test the newest version 66fbcd1 and see if that helps anything. Thanks.

**Comment by EenerNL** [comment]
Hi,

Still the same thing happening.

![image](https://github.com/sveinse/zaptec/assets/114678088/dacfefd1-3c9b-4a8a-ba96-80330631e03b)

Another thing, since the last few updates I'm experiencing another problem related to connection I think.

Previously the charger status changes very quickly from Disconnected to Waiting as soon as I connect the charging cable. Now I have to wait much longer. I'm using that status to trigger an automation for starting the charging process. Waiting is not very handy.

**Comment by sveinse** [comment]
This is so strange. All I know is that Zaptec centrally has become very slow to respond. In the newest update I increased the timeout to 10 minutes (!), so if Zaptec portal doesn't respond by then, I'd say are lost. The Zaptec status page https://zaptec.statuspage.io/ is still saying that they have performance issues. They have the note _"In the meantime, we continue to observe sporadic high levels of demand that can cause slow responsiveness. If you experience a timeout, please retry the operation."_

Perhaps I need to rethink how requests are made? Have shorter timeout, but with a retry mechanism?

> Previously the charger status changes very quickly from Disconnected to Waiting as soon as I connect the charging cable. Now I have to wait much longer. I'm using that status to trigger an automation for starting the charging process. Waiting is not very handy.

I've seen the same thing in my own installation. It is the cloud which is very slow to update the information for some unknown reason. I'll have to investigate if there is anything at all that can be done from the integration side.

**Comment by sveinse** [comment]
@EenerNL I'm still working on improving the robustness of the integration. At the same time Zaptec status (https://zaptec.statuspage.io/) now reports that the have just completed a major infrastructure upgrade. Can you test if you are still experiencing any errors?

**Comment by EenerNL** [comment]
I will monitor it the coming days. Last timeout was two hours ago (11h16).

I experienced a positive change two hours ago when I connected the car. The charger status changed immediately to waiting this time. Maybe this is a sign the connection is getting better.

**Comment by EenerNL** [comment]
Hi Sveinse,

It looks like the connection is more stable now. The last time I had statuses "unavailable" was 9th of October at 11:50h. Do you experience the same?

**Comment by sveinse** [comment]
@EenerNL things appear stable here too.

Ironically, the instability from the zaptec cloud helped me iron out a few weak points in the integration, which has become harder to test now. But... I'm not complaining. 😄

**Comment by sveinse** [comment]
Fixed in 4f9d15a960ed9c1693e22bcaa02acbad7c6f09c0 of zapted-dev. Please retest. Without know more of what metrics to expect from the Zaptec cloud, I don't think there is much more than can be done.
